From 9440a9afbdb10317cd92ed0221dad684379ca631 Mon Sep 17 00:00:00 2001
From: Galih Tama <galpt@v.recipes>
Date: Tue, 4 Aug 2026 16:43:51 +0700
Subject: [PATCH 34/46] v4.8-gpu: show the IPC boost ratio against IPC wakeups

The IPC boosts row annotated boosts as a percentage of EMA climbs,
which is structurally tiny (1.72K boosts vs 242.99K climbs reads as
0%) and the 'of tasks' label was misleading.  The funnel counters
already exist, so ratio against the IPC wakeup candidates instead:
1.72K of 5.50K candidates is 31% of IPC wakeups, a meaningful signal
for whether the boost is doing its job.

Signed-off-by: Galih Tama <galpt@v.recipes>
---
 kernel/sched/fair.c           | 1 +
 kernel/sched/infinity_sched.c | 7 ++++---
 2 files changed, 5 insertions(+), 3 deletions(-)

diff --git a/kernel/sched/fair.c b/kernel/sched/fair.c
index 28ea0d5dc..cca5e23c3 100644
--- a/kernel/sched/fair.c
+++ b/kernel/sched/fair.c
@@ -7565,6 +7565,7 @@ enqueue_task_fair(struct rq *rq, struct task_struct *p, int flags)
 			cpufreq_update_util(rq, SCHED_CPUFREQ_IOWAIT);
 
 		p->infinity.futex_waiting = false;
+		p->infinity.ipc_waiting = false;
 
 		if (p->in_iowait)
 			cpufreq_update_util(rq, SCHED_CPUFREQ_IOWAIT);
diff --git a/kernel/sched/infinity_sched.c b/kernel/sched/infinity_sched.c
index 53b874e98..5f7921cf3 100644
--- a/kernel/sched/infinity_sched.c
+++ b/kernel/sched/infinity_sched.c
@@ -294,11 +294,12 @@ static int infinity_stats_proc_handler(const struct ctl_table *ctl, int write,
 
 	cpu_rows[5].label = "IPC boosts";
 	fill_pretty_llu(cpu_rows[5].value, sizeof(cpu_rows[5].value), ipb);
-	if (emc)
+	if (ipw)
 		scnprintf(cpu_rows[5].note, sizeof(cpu_rows[5].note),
-			  "%llu%% of tasks", mul_u64_u32_div(ipb, 100, emc));
+			  "%llu%% of IPC wakeups",
+			  mul_u64_u32_div(ipb, 100, ipw));
 	else
-		strscpy(cpu_rows[5].note, "interactive IPC wakeups boosted",
+		strscpy(cpu_rows[5].note, "no IPC wakeups yet",
 			sizeof(cpu_rows[5].note));
 
 	cpu_rows[6].label = "IPC wakeups";
-- 
2.55.0

